File: cpp/3sum.cc
/*
 * Filename: 3sum.cc
 * Author: Jun Xu
 * Email: jxu.vimacs@gmail.com
 * Created Time: Thu 23 Jul 2015 08:48:43 PM EDT
 */

#include <algorithm>
#include <iostream>
#include <map>

using namespace std;

class Solution {
	public:
		vector<vector<int> > threeSum(vector<int>& nums) {
			sort(nums.begin(), nums.end());
			map<int, int> map;
			vector<vector<int> > res;

			for (vector<int>::size_type i = 0; i < nums.size(); i++) {
				map[nums[i]] = i;
			}

			for (vector<int>::size_type i = 0; i < nums.size(); i++) {
				if (i && nums[i] == nums[i-1]) {
					continue;
				}

				for (vector<int>::size_type j = i+1; j < nums.size(); j++) {
					if (j-i-1 && nums[j] == nums[j-1]) {
						continue;
					}

					int c = - nums[i] - nums[j];

					if (map.find(c) != map.end()) {
						int pc = map[c];

						if (pc > j) {
							vector<int> list;
							list.push_back(nums[i]);
							list.push_back(nums[j]);
							list.push_back(c);
							res.push_back(list);
						}
					}
				}
			}

			return res;
		}
};

File: java/3Sum.java
/*
 * Filename: 3Sum.java
 * Author: Jun Xu
 * Email: junx@bu.edu
 * Created Time: Wed 11 Mar 2015 03:56:43 PM EDT
 */

public class Solution {
	public List<List<Integer>> threeSum(int[] num) {
		Arrays.sort(num);

		List<List<Integer>> ret = new ArrayList<ArrayList<Integer>>();

		int n = num.length;
		for (int i = 0; i < n - 2; i++) {
			if (i > 0 && num[i] == num[i-1]) {
				continue;
			}

			int a = num[i];
			int start = i+1;
			int end = n-1;

			while (start < end) {
				int b = num[start];
				int c = num[end];

				if (a+b+c == 0) {
					if (start != i+1 && num[start] == num[start-1]) {
						start++;
						continue;
					}

					if (end != n-1 && num[end] == num[end+1]) {
						end--;
						continue;
					}

					List<Integer> tuple = new ArrayList<Integer>();
					tuple.add(a);
					tuple.add(b);
					tuple.add(c);
					ret.add(tuple);
					start++;
					end--;
				} else if (a+b+c > 0) {
					end--;
				} else {
					start++;
				}
			}
		}

		return ret;
	}
}
